File: src/grammarFixer.js
/**
 * Grammar Fixer Module
 * 
 * This module provides functionality to fix grammar in text using Ollama and Gemma3.
 * It returns structured JSON objects detailing each grammar change.
 */

const { Ollama } = require('ollama');

/**
 * Fixes grammar in the provided text using Ollama and Gemma3
 * 
 * @param {string} text - The text to analyze and correct
 * @param {Object} options - Optional configuration
 * @param {string} options.model - Model to use (default: 'gemma3')
 * @param {string} options.host - Ollama host (default: 'http://localhost:11434')
 * @returns {Promise<Array>} Array of correction objects with structure:
 *   {
 *     location: { start: number, end: number },
 *     oldText: string,
 *     newText: string
 *   }
 */
async function fixGrammar(text, options = {}) {
  if (!text || typeof text !== 'string') {
    throw new Error('Invalid input: text must be a non-empty string');
  }

  const config = {
    model: options.model || 'gemma3',
    host: options.host || 'http://localhost:11434'
  };
  
  try {
    // Call Ollama with Gemma3 to analyze the text
    const corrections = await analyzeTextWithOllama(text, config);
    
    // Process and format the corrections
    const formattedCorrections = processCorrections(text, corrections);
    
    return formattedCorrections;
  } catch (error) {
    console.error('Error fixing grammar:', error);
    throw error;
  }
}

/**
 * Analyzes text using Ollama and Gemma3 model
 * 
 * @private
 * @param {string} text - The text to analyze
 * @param {Object} config - Configuration object
 * @returns {Promise<Array>} Raw corrections from Ollama
 */
async function analyzeTextWithOllama(text, config) {
  const ollama = new Ollama({ host: config.host });
  
  // Create a detailed prompt for grammar correction
  const prompt = `You are a grammar correction assistant. Analyze the following text and identify ALL grammar errors.

For each error, provide the response in STRICT JSON format as an array of objects. Each object must have:
- "oldText": the exact incorrect text from the original
- "newText": the corrected text
- "explanation": brief explanation of the error

Original text: "${text}"

Return ONLY the JSON array, nothing else. If there are no errors, return an empty array [].

Example format:
[
  {
    "oldText": "dont",
    "newText": "doesn't",
    "explanation": "Incorrect contraction"
  }
]`;

  try {
    const response = await ollama.generate({
      model: config.model,
      prompt: prompt,
      stream: false,
      options: {
        temperature: 0.3,
        top_p: 0.9
      }
    });

    return parseOllamaResponse(response.response, text);
  } catch (error) {
    // If Ollama is not available, throw a more descriptive error
    if (error.code === 'ECONNREFUSED' || 
        error.cause?.code === 'ECONNREFUSED' ||
        error.message?.includes('connect') ||
        error.message?.includes('fetch failed')) {
      throw new Error('Unable to connect to Ollama. Please ensure Ollama is running (ollama serve)');
    }
    throw error;
  }
}

/**
 * Parses the response from Ollama to extract corrections
 * 
 * @private
 * @param {string} responseText - The response text from Ollama
 * @param {string} originalText - The original text for validation
 * @returns {Array} Array of correction objects
 */
function parseOllamaResponse(responseText, originalText) {
  try {
    // Extract JSON from the response (it might have extra text)
    let jsonMatch = responseText.match(/\[[\s\S]*\]/);
    if (!jsonMatch) {
      console.warn('No JSON array found in Ollama response, returning empty corrections');
      return [];
    }

    const corrections = JSON.parse(jsonMatch[0]);
    
    if (!Array.isArray(corrections)) {
      console.warn('Ollama response is not an array, returning empty corrections');
      return [];
    }

    // Validate and enrich corrections with position information
    return corrections.filter(correction => {
      return correction.oldText && 
             correction.newText && 
             typeof correction.oldText === 'string' &&
             typeof correction.newText === 'string';
    });
  } catch (error) {
    console.warn('Failed to parse Ollama response:', error.message);
    return [];
  }
}

/**
 * Processes raw corrections into the required format
 * 
 * @private
 * @param {string} originalText - The original text
 * @param {Array} rawCorrections - Raw corrections from Ollama
 * @returns {Array} Formatted correction objects
 */
function processCorrections(originalText, rawCorrections) {
  const formattedCorrections = [];
  let searchStartIndex = 0;
  
  for (const correction of rawCorrections) {
    // Find the position of the old text in the original
    const position = findSubstringPosition(originalText, correction.oldText, searchStartIndex);
    
    if (position) {
      const formattedCorrection = {
        location: {
          start: position.start,
          end: position.end
        },
        oldText: correction.oldText,
        newText: correction.newText
      };
      
      // Add explanation if available
      if (correction.explanation) {
        formattedCorrection.explanation = correction.explanation;
      }
      
      // Validate before adding
      if (isValidCorrection(formattedCorrection)) {
        formattedCorrections.push(formattedCorrection);
        // Move search index forward to handle multiple occurrences
        searchStartIndex = position.end;
      }
    } else {
      console.warn(`Could not find "${correction.oldText}" in original text`);
    }
  }
  
  return formattedCorrections;
}

/**
 * Helper function to find the position of a substring in text
 * 
 * @private
 * @param {string} text - The text to search in
 * @param {string} substring - The substring to find
 * @param {number} startIndex - Starting index for search
 * @returns {Object} Object with start and end positions
 */
function findSubstringPosition(text, substring, startIndex = 0) {
  const start = text.indexOf(substring, startIndex);
  if (start === -1) {
    return null;
  }
  
  return {
    start,
    end: start + substring.length
  };
}

/**
 * Validates a correction object
 * 
 * @private
 * @param {Object} correction - The correction object to validate
 * @returns {boolean} True if valid, false otherwise
 */
function isValidCorrection(correction) {
  if (!correction || typeof correction !== 'object') {
    return false;
  }
  
  if (!correction.location || typeof correction.location !== 'object') {
    return false;
  }
  
  if (typeof correction.location.start !== 'number' || typeof correction.location.end !== 'number') {
    return false;
  }
  
  if (typeof correction.oldText !== 'string' || typeof correction.newText !== 'string') {
    return false;
  }
  
  return true;
}

module.exports = {
  fixGrammar,
  // Export helper functions for testing
  isValidCorrection,
  findSubstringPosition
};

File: src/grammar_fixer.py
"""
Grammar Fixer Module

This module provides functionality to fix grammar in text using Ollama and Gemma3.
It returns structured JSON objects detailing each grammar change.
"""

import json
from typing import List, Dict, Optional, Any


class GrammarFixer:
    """
    Grammar Fixer class for analyzing and correcting grammar using Ollama and Gemma3.
    """

    def __init__(self):
        """Initialize the GrammarFixer."""
        pass

    async def fix_grammar(self, text: str) -> List[Dict[str, Any]]:
        """
        Fixes grammar in the provided text using Ollama and Gemma3.

        Args:
            text: The text to analyze and correct

        Returns:
            List of correction objects with structure:
            {
                "location": {"start": int, "end": int},
                "oldText": str,
                "newText": str
            }

        Raises:
            ValueError: If text is invalid (None, empty, or not a string)
        """
        if not text or not isinstance(text, str):
            raise ValueError("Invalid input: text must be a non-empty string")

        try:
            # Placeholder: In actual implementation, this will call Ollama with Gemma3
            corrections = await self._analyze_text_with_ollama(text)

            # Process and format the corrections
            formatted_corrections = self._process_corrections(text, corrections)

            return formatted_corrections
        except ValueError:
            raise
        except Exception as error:
            print(f"Error fixing grammar: {error}")
            raise

    async def _analyze_text_with_ollama(self, text: str) -> List[Dict[str, Any]]:
        """
        Analyzes text using Ollama and Gemma3 model.

        Args:
            text: The text to analyze

        Returns:
            Raw corrections from Ollama

        Note:
            This is a placeholder implementation. In production, this would:
            - Import ollama library
            - Make API call to Ollama with Gemma3 model
            - Parse and return the response
        """
        # TODO: Implement Ollama API call
        # Example implementation structure:
        """
        import ollama
        
        response = await ollama.chat(
            model='gemma3',
            messages=[{
                'role': 'user',
                'content': f'Fix the grammar in this text and provide changes in JSON format: "{text}"'
            }],
        )
        
        return self._parse_ollama_response(response)
        """

        # Placeholder: Return empty array for now
        print(f"Analyzing text with Ollama/Gemma3 (placeholder): {text}")
        return []

    def _process_corrections(
        self, original_text: str, raw_corrections: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """
        Processes raw corrections into the required format.

        Args:
            original_text: The original text
            raw_corrections: Raw corrections from Ollama

        Returns:
            Formatted correction objects

        Note:
            This function should:
            1. Parse the corrections from Ollama's response
            2. Calculate exact character positions (start, end)
            3. Extract old and new text
            4. Return formatted array of correction objects
        """
        formatted_corrections = []

        # Placeholder implementation
        for correction in raw_corrections:
            formatted_corrections.append(
                {
                    "location": {
                        "start": correction.get("start", 0),
                        "end": correction.get("end", 0),
                    },
                    "oldText": correction.get("oldText", ""),
                    "newText": correction.get("newText", ""),
                }
            )

        return formatted_corrections

    @staticmethod
    def find_substring_position(
        text: str, substring: str, start_index: int = 0
    ) -> Optional[Dict[str, int]]:
        """
        Helper function to find the position of a substring in text.

        Args:
            text: The text to search in
            substring: The substring to find
            start_index: Starting index for search (default: 0)

        Returns:
            Dictionary with start and end positions, or None if not found
        """
        start = text.find(substring, start_index)
        if start == -1:
            return None

        return {"start": start, "end": start + len(substring)}

    @staticmethod
    def is_valid_correction(correction: Any) -> bool:
        """
        Validates a correction object.

        Args:
            correction: The correction object to validate

        Returns:
            True if valid, False otherwise
        """
        if not correction or not isinstance(correction, dict):
            return False

        if "location" not in correction or not isinstance(correction["location"], dict):
            return False

        location = correction["location"]
        if not isinstance(location.get("start"), int) or not isinstance(
            location.get("end"), int
        ):
            return False

        if not isinstance(correction.get("oldText"), str) or not isinstance(
            correction.get("newText"), str
        ):
            return False

        return True


def fix_grammar(text: str) -> List[Dict[str, Any]]:
    """
    Convenience function to fix grammar in text.

    Args:
        text: The text to analyze and correct

    Returns:
        List of correction objects in JSON format

    Note:
        This is a synchronous wrapper. For async usage, use GrammarFixer class directly.
    """
    import asyncio

    fixer = GrammarFixer()
    return asyncio.run(fixer.fix_grammar(text))


# Export functions for API compatibility
__all__ = ["GrammarFixer", "fix_grammar"]
